lib/IDE/CompletionLookup.cpp

```diff
@@ -207,6 +161,52 @@ bool swift::ide::isCompletionDeclContextLocalContext(DeclContext *DC) {
   return true;
 }
 
+/// Returns \c true if \p DC can handles async call.
+bool swift::ide::canDeclContextHandleAsync(const DeclContext *DC) {
+  if (auto *func = dyn_cast<AbstractFunctionDecl>(DC))
+    return func->isAsyncContext();
+
+  if (auto *closure = dyn_cast<ClosureExpr>(DC)) {
+    // See if the closure has 'async' function type.
+    if (auto closureType = closure->getType())
+      if (auto fnType = closureType->getAs<AnyFunctionType>())
+        if (fnType->isAsync())
+          return true;
+
+    // If the closure doesn't contain any async call in the body, closure itself
+    // doesn't have 'async' type even if 'async' closure is expected.
+    //   func foo(fn: () async -> Void)
+    //   foo { <HERE> }
+    // In this case, the closure is wrapped with a 'FunctionConversionExpr'
+    // which has 'async' function type.
```

Comment:
This assume a type-checked version of the AST, is that always the case?

Reply:
I believe so. The call side inside `CompletionLookup` is intended for the legacy code completion, which works on a type-checked AST and https://github.com/apple/swift/pull/41885/files/03d819f442f05b58f6a2b1816b2e280292db5329#diff-3088e637c4f06d0fc6cb38370068cd6fa6fe8dcaa10ec35c9086602c59eaba17R149 works on the already type-checked part of the AST to determine async-ness after failing to determine async-ness from the solution.